torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -234,8 +236,64 @@ inline void errorIfCapturingNonCapturableNCCL() {
   }
 }
 
+void markRankNotStuck(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x00;
+}
+
+void markRankStuck(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x01;
+}
+
+void markRankTimedOut(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x02;
+}
+
 } // namespace
 
+RankStatusHelper::RankStatusHelper(size_t worldSize): worldSize_(worldSize) {
+  // assume file exists
+  int fd;
+  if ((fd = open("rank_statuses", O_RDWR, 0)) == -1) {
+    LOG(ERROR) << "unable to open rank_statuses";
+  }
+  handle_ = static_cast<char*>(mmap(
+                        nullptr,
+                        worldSize,
```

Comment:
The PR description says that the file would have n bytes where n is the total number of ranks on a node. However `worldSize` is actually the global world_size, so if you have 2 nodes with 8 gpus each resulting in a total of 16 ranks, then worldSize will be 16.

I think we can just use the number of GPUs here as the file size instead of worldSize